proposals/index-range-changes.md

```diff
@@ -0,0 +1,297 @@
+in existing code without fear of lost performance.
+
+This proposal specifically does not want to change how `Index` and `Range` type binding occurs. The index and range
+expressions continue to have the same syntax and types.
+
+## Detailed Design 
+### Countable Types
+Any type which has a property named `Length` or `Count` with an accessible getter and a return type of `int` is 
+considered Countable. The language can make use of this property to convert an expression of type `Index` into an `int`
+at the point of the expression without the need to use the type `Index` at all. In case both `Length` and `Count`
+are present, `Length` will be preferred. 
+
+For simplicity going forward, the proposal will use the name `Length` to represent `Count` or `Length`.
+
+### Index and Range implementations are known
+The implementations of `Index` and `Range` are considered to be known and side effect free. Much like 
+`ValueTuple<T1, T2>`, the language can assume a standard implementation and emit code inline which represents the 
+implementation of methods on `Index` and `Range`. This implementation includes methods like `GetOffset` or conversions
+like the implicit conversion from `int` to `Index`.
+
+All arithmetic operations will be emitted using an `unchecked` context. That matches the context
+in which `Index` and `Range` are compiled in.
+
+### Implicit Index support
+The language will provide an instance indexer member with a single parameter of type `Index` for types which meet the 
+following criteria:
+
+- The type is Countable.
+- The type has an accessible instance indexer which takes a single `int` as the argument.
+- The type does not have an accessible instance indexer which takes a single `Index` as the parameter.
+
+For such types, the language will act as if there is an index member of the form `T this[Index index]` where `T` is 
+the return type of the `int` based indexer including any `ref` style annotations. The new member will have the same 
+`get` and `set` members with matching accessibilty as the `int` indexer. 
+
+The new indexer will be implemented by converting the argument of type `Index` into an `int` and emitting a call 
+to the `int` based indexer. For discussion purposes, lets use the example of `receiver[expr]`. The conversion of 
+`expr` to `int` will occur as follows:
+
+- When the argument is of the form `^expr2` and the type of `expr2` is `int`, it will be translated to 
+`receiver.Length - expr2`.
+- Otherwise, it will be translated as `expr.GetOffset(receiver.Length)`.
+
+This allows for developers to use the `Index` feature on existing types without the need for modification. For example:
+
+``` csharp
+List<char> list = ...;
+var value = list[^1]; 
+
+// Gets translated to 
+var value = list[list.Length - 1]; 
+```
+
+The `receiver` and `Length` expressions will be spilled as appropriate to ensure any side effects are only executed 
+once. For example:
+
+``` csharp
+class Collection {
```

Comment:
In [corefx/coding-style.md at master · dotnet/corefx](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md ) NO.1, we use Allman style braces, where each brace begins on a new line. A single line statement block can go without braces but the block must be properly indented on its own line and must not be nested in other statement blocks that use braces (See issue 381 for examples). One exception is that a using statement is permitted to be nested within another using statement by starting on the following line at the same indentation level, even if the nested using contains a controlled block.